### PR TITLE
[DPE-3596] Test app improvements

### PR DESCRIPTION
**Pull request by dragomirp**
* Add show action for continuous writes
* Add database relation
* Add sleep interval to the continuous writes loop
* Make `run-sql` action find relation-id on its own

### CODE REVIEW

File: metadata.yaml [hunk 1]
```diff
@@ -19,10 +19,14 @@ maintainers:
 
 
 requires:
+  database:
+    interface: postgresql_client
+    limit: 1
```
Comment (dragomirp) [inline metadata.yaml:24]: Need to switch all the clients, before we can remove `first-database`

File: src/charm.py [hunk 2]
```diff
@@ -237,10 +291,15 @@ def _on_cluster2_endpoints_changed(self, event: DatabaseEndpointsChangedEvent) -
     @property
     def _connection_string(self) -> Optional[str]:
         """Returns the PostgreSQL connection string."""
-        data = list(self.first_database.fetch_relation_data().values())[0]
+        db_data = list(self.database.fetch_relation_data().values())
+        if db_data:
+            data = db_data[0]
+        else:
+            data = list(self.first_database.fetch_relation_data().values())[0]
```
Comment (dragomirp) [inline src/charm.py:298]: Use the new rel for continuous writes.

File: src/charm.py [hunk 3]
```diff
@@ -400,16 +477,17 @@ def _on_run_sql_action(self, event: ActionEvent):
...
             relation = self.first_database
+        if relation_name == self.database.relation_name:
+            relation = self.database
         elif relation_name == self.second_database.relation_name:
             relation = self.second_database
         else:
             event.fail(message="invalid relation name")
 
-        databag = relation.fetch_relation_data()[relation_id]
+        databag = list(relation.fetch_relation_data().values())[0]
```
Comment (dragomirp) [inline src/charm.py:490]: This removes the need for passing the relation-id, but then we can only have one relation per interface.

File: src/charm.py [hunk 4]
```diff
@@ -184,6 +207,37 @@
         count = self._count_writes()
         self._start_continuous_writes(count + 1)
 
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application."""
+        # Retrieve the credentials using the charm library.
+        logger.info(f"database credentials: {event.username} {event.password}")
```
[the remark this answers is not in the document]
Comment (dragomirp) [reply src/charm.py:213]: Suppressing the security check.

File: src/charm.py [hunk 5]
```diff
@@ -323,6 +382,23 @@ def _on_start_continuous_writes_action(self, event: ActionEvent) -> None:
         self._start_continuous_writes(1)
         event.set_results({"result": "True"})
 
+    def _on_show_continuous_writes_action(self, event: ActionEvent) -> None:
+        """Count the continuous writes."""
+        try:
+            # Create the table to write records on and also a unique index to prevent duplicate
```
Comment (taurus-forever) [inline src/charm.py:388]: Copy&Paste?

File: config.yaml [hunk 6]
```diff
@@ -8,3 +8,8 @@ options:
     description: |
       Request roles instead of extensions
     type: boolean
+  sleep_interval:
+    default: 0
+    description: |
+      How long to sleep between writes
```
Comment (taurus-forever) [inline config.yaml:14]: Please indicate the unit: seconds, minutes, milliseconds? Tnx!

**Review by taurus-forever — APPROVED** [review]
LGTM! Two minor nitpicking, otherwise LGTM.

File: src/charm.py (hunk 5 above)
Comment (dragomirp) [reply src/charm.py]: Removed the comment

File: config.yaml (hunk 6 above)
Comment (dragomirp) [reply config.yaml]: Added unit

File: src/continuous_writes.py [hunk 7]
```diff
@@ -80,7 +82,11 @@ def write(write_value: int) -> None:
 def main():
     """Main executor."""
     starting_number = int(sys.argv[1])
-    continuous_writes(starting_number)
+    if len(sys.argv) > 2:
+        sleep_interval = int(sys.argv[2]) / 1000
+    else:
+        sleep_interval = 0.0
```
Comment (dragomirp) [inline src/continuous_writes.py:88]: Sleep for ms

**Review by taurus-forever — APPROVED** [review]
Thank you!

**Review by lucasgameiroborges — APPROVED** [review]
Nice!

**Pull request merged**